File: android/src/main/java/com/flutter_tim_plugin/MessageFactory.java
package com.flutter_tim_plugin;

import android.util.JsonWriter;


import com.tencent.imsdk.TIMCallBack;
import com.tencent.imsdk.TIMCustomElem;
import com.tencent.imsdk.TIMElem;
import com.tencent.imsdk.TIMElemType;
import com.tencent.imsdk.TIMFileElem;
import com.tencent.imsdk.TIMImage;
import com.tencent.imsdk.TIMImageElem;
import com.tencent.imsdk.TIMMessage;
import com.tencent.imsdk.TIMSoundElem;
import com.tencent.imsdk.TIMTextElem;
import com.tencent.imsdk.TIMVideoElem;

import org.json.JSONObject;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.JSONUtil;

public class MessageFactory {

    private MessageFactory() {
    }

    private static class SingletonInstance {

        private static final MessageFactory INSTANCE = new MessageFactory();
    }

    public static MessageFactory getInstance() {
        return SingletonInstance.INSTANCE;
    }






    public String basicMessage2String(TIMMessage msg){
        Map map = buildBasicMap(msg);
        List elementList=new ArrayList();
        for (int i = 0; i < msg.getElementCount(); i++) {
            TIMElem element =  msg.getElement(i);
            Map elementMap=new HashMap();
            elementMap.put("type",element.getType().value());

            elementList.add(elementMap);

        }
        map.put("elementList",elementList);
        return new JSONObject(map).toString();
    }

    public String imageMessage2String(TIMMessage msg) {

        Map dataMap=buildBasicMap(msg);


        List elementList=new ArrayList();
        dataMap.put("messageType",TIMElemType.Image.value());
        for (int i = 0; i < msg.getElementCount(); i++) {
            TIMImageElem element = (TIMImageElem) msg.getElement(i);
            Map elementMap=new HashMap();
            elementMap.put("path",element.getPath());
            elementMap.put("imageFormat",element.getImageFormat());
            elementMap.put("level",element.getLevel());
            elementMap.put("taskId",element.getTaskId());


            ArrayList imageList = new ArrayList();
            for (int j = 0; j < element.getImageList().size(); j++) {
                Map imageListMap=new HashMap();
                imageListMap.put("url",element.getImageList().get(j).getUrl());
                imageListMap.put("height",element.getImageList().get(j).getHeight());
                imageListMap.put("size",element.getImageList().get(j).getSize());
                imageListMap.put("type",element.getImageList().get(j).getType().value());
                imageListMap.put("uuid",element.getImageList().get(j).getUuid());
                imageListMap.put("width",element.getImageList().get(j).getWidth());

                imageList.add(imageListMap);
            }
            elementMap.put("imageList",imageList);


            elementList.add(elementMap);


        }

        dataMap.put("elementList",elementList);

        System.out.println("打印  "+new JSONObject(dataMap).toString());

        return new JSONObject(dataMap).toString();
    }

    public String soundMessage2String(TIMMessage msg) {
        Map dataMap=buildBasicMap(msg);

        dataMap.put("messageType",TIMElemType.Sound.value());

        List elementList=new ArrayList();
        System.out.println( "发送消息成功  getElementCount  "+msg.getElementCount());
        for (int i = 0; i < msg.getElementCount(); i++) {
            TIMSoundElem element = (TIMSoundElem) msg.getElement(i);
            Map elementMap=new HashMap();
            elementMap.put("path",element.getPath());
            elementMap.put("duration",element.getDuration());
            elementMap.put("dataSize",element.getDataSize());
            elementMap.put("taskId",element.getTaskId());
            elementMap.put("uuid",element.getUuid());


            elementList.add(elementMap);

        }

        dataMap.put("elementList",elementList);

        System.out.println("打印  "+new JSONObject(dataMap).toString());

        return new JSONObject(dataMap).toString();
    }


    public String fileMessage2String(TIMMessage msg){



        Map dataMap=buildBasicMap(msg);
        dataMap.put("messageType",TIMElemType.File.value());


        List elementList=new ArrayList();
        System.out.println( "收到文件消息成功  getElementCount  "+msg.getElementCount());
        for (int i = 0; i < msg.getElementCount(); i++) {
            TIMFileElem element = (TIMFileElem) msg.getElement(i);
            Map elementMap=new HashMap();
            elementMap.put("fileName",element.getFileName());
            elementMap.put("fileSize",element.getFileSize());
            elementMap.put("path",element.getPath());
            elementMap.put("taskId",element.getTaskId());
            elementMap.put("uuid",element.getUuid());


            System.out.println("打印   "+msg.getMessageLocator().toString());

            elementList.add(elementMap);

        }

        dataMap.put("elementList",elementList);

        System.out.println("打印  "+new JSONObject(dataMap).toString());

        return new JSONObject(dataMap).toString();
    }

    public String videoMessage2String(TIMMessage msg) {
        Map dataMap=buildBasicMap(msg);

        dataMap.put("messageType",TIMElemType.Video.value());

        List elementList=new ArrayList();
        System.out.println( "消息视频  getElementCount  "+msg.getElementCount());
        for (int i = 0; i < msg.getElementCount(); i++) {
            TIMVideoElem element = (TIMVideoElem) msg.getElement(i);

            System.out.println(" 视频信息  "+element.getVideoPath());
            Map elementMap=new HashMap();
            elementMap.put("videoPath",element.getVideoPath());
            elementMap.put("snapshotPath",element.getSnapshotPath());

            elementMap.put("videoUUID",element.getVideoInfo().getUuid());
            elementMap.put("videoType",element.getVideoInfo().getType());
            elementMap.put("videoSize",element.getVideoInfo().getSize());
            elementMap.put("duaration",element.getVideoInfo().getDuaration());

            elementMap.put("snapshotUUID",element.getSnapshotInfo().getUuid());
            elementMap.put("snapshotType",element.getSnapshotInfo().getType());
            elementMap.put("snapshotHeight",element.getSnapshotInfo().getHeight());
            elementMap.put("snapshotWidth",element.getSnapshotInfo().getWidth());


            elementList.add(elementMap);

        }

        dataMap.put("elementList",elementList);

        System.out.println("打印  "+new JSONObject(dataMap).toString());

        return new JSONObject(dataMap).toString();
    }

    private Map buildBasicMap(TIMMessage msg){
        Map dataMap=new HashMap();
        dataMap.put("msgId",msg.getMsgId());
        dataMap.put("msgSeq",msg.getSeq());
        dataMap.put("rand",msg.getRand());
        dataMap.put("time",msg.timestamp());
        dataMap.put("isSelf",msg.isSelf());
        dataMap.put("isPeerReaded",msg.isPeerReaded());
        dataMap.put("isRead",msg.isRead());
        dataMap.put("status",msg.status().getStatus());
        dataMap.put("sender",msg.getSender());
        return dataMap;
    }
    public String customMessage2String(TIMMessage msg) {
        Map dataMap=buildBasicMap(msg);

        dataMap.put("messageType",TIMElemType.Custom.value());

        List elementList=new ArrayList();
        System.out.println( "发送消息成功  getElementCount  "+msg.getElementCount());
        for (int i = 0; i < msg.getElementCount(); i++) {
            TIMCustomElem element = (TIMCustomElem) msg.getElement(i);
            Map elementMap=new HashMap();
            elementMap.put("data",element.getData());
            elementList.add(elementMap);

        }

        dataMap.put("elementList",elementList);

        System.out.println("打印  "+new JSONObject(dataMap).toString());

        return new JSONObject(dataMap).toString();
    }
    public String textMessage2String(TIMMessage msg) {
        Map dataMap=buildBasicMap(msg);

        dataMap.put("messageType",TIMElemType.Text.value());

        List elementList=new ArrayList();

        for (int i = 0; i < msg.getElementCount(); i++) {
            TIMTextElem element = (TIMTextElem) msg.getElement(i);
            Map elementMap=new HashMap();
            elementMap.put("text",element.getText());
            elementList.add(elementMap);

        }

        dataMap.put("elementList",elementList);

        System.out.println("打印  "+new JSONObject(dataMap).toString());

        return new JSONObject(dataMap).toString();
    }

    public List<String> message2List(List<TIMMessage> timList){
        List<String> list=new ArrayList<>();
        for (TIMMessage msg : timList) {

            TIMElem elem = msg.getElement(0);

            //获取当前元素的类型
            TIMElemType elemType = elem.getType();
            System.out.println( "elem type: " + elemType.name());
            if (elemType == TIMElemType.Text) {

                list.add(textMessage2String(msg));
                //处理文本消息
            } else if (elemType == TIMElemType.Image) {

                list.add(imageMessage2String(msg));
            }else if (elemType == TIMElemType.Sound) {

                list.add(soundMessage2String(msg));
            }else if (elemType == TIMElemType.File) {

                list.add(fileMessage2String(msg));
            }else if (elemType == TIMElemType.Video) {

                list.add(videoMessage2String(msg));
            }else if (elemType == TIMElemType.Custom) {

                list.add(customMessage2String(msg));
            }else {



                list.add(basicMessage2String(msg));
                System.out.println("message2List      elemType     "+elemType);
            }

        }


        return list;
    }



}

File: ios/Classes/MessageFactory.swift
//
//  MessageFactory.swift
//  flutter_tim_plugin
//
//  Created by kbz on 2019/12/29.
//

class MessageFactory{
    
    static func  buildBasicMap(_ msg : TIMMessage) -> [String:Any]{
        var dataMap =  [String:Any]();
        dataMap["msgId"] = msg.msgId();

        dataMap["msgSeq"] = msg.locator()?.seq;
        dataMap["rand"] = msg.locator()?.rand;
        dataMap["time"] = Int(msg.timestamp()!.timeIntervalSince1970);
        dataMap["isSelf"] = msg.isSelf();
        dataMap["isPeerReaded"] = msg.isPeerReaded();
        dataMap["isRead"] = msg.isReaded();
        dataMap["status"] = msg.status().rawValue;
        dataMap["sender"] = msg.sender();
       
        return dataMap;
    }
    
    
    static func customMessage2String(_ msg : TIMMessage)-> String{
        
        var map = buildBasicMap(msg);
        
        map["messageType"] = MessageType.Custom;
        
        var elementList = Array<[String:Any]>();
              
        var count = msg.elemCount();
        for index in 0..<count{
            var element  =  msg.getElem(index) as! TIMCustomElem;
            
       
            var elementMap = ["data":Array(element.data)]
            elementList.append(elementMap);
            
        }
        map["elementList"] = elementList;
        return convertDictionaryToJSONString(dict: map)
    }
    
    static func fileMessage2String(_ msg : TIMMessage)-> String{
         var map = buildBasicMap(msg);
         
         map["messageType"] = MessageType.Custom;
         
         var elementList = Array<[String:Any]>();
               
         var count = msg.elemCount();
         for index in 0..<count{
             var element  =  msg.getElem(index) as! TIMFileElem;
             
        
            var elementMap = [String:Any]();
            elementMap["fileName"] = element.filename;
            elementMap["fileSize"] = element.fileSize;
            elementMap["path"] = element.path;
            elementMap["taskId"] = element.taskId;
            elementMap["uuid"] = element.uuid;
            
             elementList.append(elementMap);
             
         }
         map["elementList"] = elementList;
         return convertDictionaryToJSONString(dict: map)
    }
    
    static func videoMessage2String(_ msg : TIMMessage)-> String{
         var map = buildBasicMap(msg);
         
         map["messageType"] = MessageType.Sound;
         
         var elementList = Array<[String:Any]>();
               
         var count = msg.elemCount();
         for index in 0..<count{
             var element  =  msg.getElem(index) as! TIMVideoElem;
             
        
            var elementMap = [String:Any]();
            elementMap["videoPath"] = element.videoPath;
            elementMap["snapshotPath"] = element.snapshotPath;
            elementMap["videoUUID"] = element.video.uuid;
            elementMap["videoType"] = element.video.type;
            elementMap["videoSize"] = element.video.size;
            elementMap["duaration"] = element.video.duration;
            
            elementMap["snapshotUUID"] = element.snapshot.uuid;
            elementMap["snapshotType"] = element.snapshot.type;
            elementMap["snapshotHeight"] = element.snapshot.height;
            elementMap["snapshotWidth"] = element.snapshot.width;
            


            elementList.append(elementMap);
             
         }
         map["elementList"] = elementList;
         return convertDictionaryToJSONString(dict: map)
    }
    
    static func soundMessage2String(_ msg : TIMMessage)-> String{

         var map = buildBasicMap(msg);
         
         map["messageType"] = MessageType.Sound;
         
         var elementList = Array<[String:Any]>();
               
         var count = msg.elemCount();
         for index in 0..<count{
             var element  =  msg.getElem(index) as! TIMSoundElem;
             
        
             var elementMap = [String:Any]();
            elementMap["path"] = element.path;
            elementMap["duration"] = element.second;
            elementMap["dataSize"] = element.dataSize;
            elementMap["taskId"] = element.taskId;
            elementMap["uuid"] = element.uuid;
            
            
        
            elementList.append(elementMap);
             
         }
         map["elementList"] = elementList;
         return convertDictionaryToJSONString(dict: map)
    }
    
    static func imageMessage2String(_ msg : TIMMessage)-> String{
     
            var map = buildBasicMap(msg);
            
            map["messageType"] = MessageType.Image;
            
            var elementList = Array<[String:Any]>();
                  
            var count = msg.elemCount();
            for index in 0..<count{
                
       
                var element  =  msg.getElem(index) as! TIMImageElem;
                
           
                var elementMap = [String:Any]();
                elementMap["path"] = element.path;
                elementMap["imageFormat"] = element.format.rawValue;
                elementMap["level"] = element.level.rawValue;
                elementMap["taskId"] = element.taskId;
        
                
                var images = Array<Any>();
                
                var timImageList = element.imageList as! Array<TIMImage>;
                for image in timImageList {
                    var imageListMap = [String:Any]();
                    
                    imageListMap["url"] = image.url;
                    imageListMap["height"] = image.height;
                    imageListMap["size"] = image.size;
                    imageListMap["type"] = image.type.rawValue;
                    imageListMap["uuid"] = image.uuid;
                    imageListMap["width"] = image.width;
                    
                    images.append(imageListMap);
                }
                
        
                elementMap["imageList"] = images;
                elementList.append(elementMap);
                
            }
            map["elementList"] = elementList;
            return convertDictionaryToJSONString(dict: map)
    }
    
    static func textMessage2String(_ msg : TIMMessage)-> String{
        var map = buildBasicMap(msg);
          
          map["messageType"] = MessageType.Text;
          
          var elementList = Array<[String:Any]>();
                var count = msg.elemCount();
                for index in 0..<count{
                    var element  =  msg.getElem(index) as! TIMTextElem;
                         
                    
                    var elementMap = [String:Any]();
    
                    elementMap["text"] = element.text;
                    elementList.append(elementMap);
                         
                     }
        map["text"] = elementList;
        return convertDictionaryToJSONString(dict: map)
    }
    
    static func  basicMessage2String(_ msg : TIMMessage)-> String{
        var map = buildBasicMap(msg);
        var elementList = Array<[String:Any]>();
        
        var count = msg.elemCount();

  
        for index in 0..<count{
            var element : TIMElem =  msg.getElem(index)
            var elementMap = ["type":getElemType(element)]
            
            elementList.append(elementMap);
        }
        
       
        
        map["elementList"] = elementList;

        return convertDictionaryToJSONString(dict: map)


    }
    static func message2List(timList : Array<TIMMessage>)-> Array<String>{
        var list = Array<String>();
        
    
        for msg in timList {
            let elmCount = msg.elemCount();

            let elemType = getElemType(msg.getElem(0));
            
     
            if (elemType == MessageType.Text) {
   
                list.append(textMessage2String(msg))
              }else if (elemType == MessageType.Image) {

                list.append(imageMessage2String(msg));
              }else if (elemType == MessageType.Sound) {

                  list.append(soundMessage2String(msg));
              }else if (elemType == MessageType.File) {

                  list.append(fileMessage2String(msg));
              }else if (elemType == MessageType.Video) {

                  list.append(videoMessage2String(msg));
              }else if (elemType == MessageType.Custom) {

                  list.append(customMessage2String(msg));
              }else {



                  list.append(basicMessage2String(msg));
                  
              }
            
            

    
            
        }

        return list;
        
    }
    
    
    static func getElemType(_ elem : TIMElem)-> Int{
        if(elem is TIMTextElem){
            return MessageType.Text
        }else if(elem is TIMFaceElem){
            return MessageType.Face
        }else if(elem is TIMLocationElem){
            return MessageType.Location
        }else if(elem is TIMVideoElem){
            return MessageType.Video
        }else if(elem is TIMSoundElem){
            return MessageType.Sound
        }else if(elem is TIMFileElem){
            return MessageType.File
        }else if(elem is TIMCustomElem){
            return MessageType.Custom
        }else if(elem is TIMImageElem){
            return MessageType.Image
        }else{
            return -99
        }
    }
    static func convertDictionaryToJSONString(dict : Dictionary<String,Any>)->String {
        let data = try? JSONSerialization.data(withJSONObject: dict, options: JSONSerialization.WritingOptions.init(rawValue: 0))
        let jsonStr = String(data: data!, encoding: String.Encoding(rawValue: String.Encoding.utf8.rawValue))
        return jsonStr! as String
    }


}
